parlai/core/torch_agent.py

```diff
@@ -1055,14 +1065,15 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None) -> bool:
                         f'list:\n{compatible_list}'
                     )
                 self.optimizer = MemoryEfficientFP16Optimizer(self.optimizer)
-        # TODO: we might want to hard reset optimizers here in the
-        # case of fine tuning. Some rudimentary experiments seemed to
-        # indicate that keeping adam weights around was desirable, so this
-        # will remain the behavior for the time being.
+
+        if is_finetune:
+            logging.warning('Detected a fine-tune run. Resetting the optimizer.')
+            return True
+
         if optim_states and saved_optim_type != opt['optimizer']:
             # we changed from adam to adamax, or sgd to adam, or similar
             logging.warning('Not loading optim state since optim class changed.')
-            return False
+            return True
         elif optim_states:
             # check for any fp16/fp32 conversions we need to do
             optimstate_fp16 = 'loss_scaler' in optim_states
```

Comment:
i can't leave a comment for it, but are the semantics correct for line 1099? the `elif not optimstate_fp16 and self.fp16` block? are we returning `True` always because of the lower precision conversion?

Reply:
agreed